### A compilation of code I wrote for my Systems Biology of Disease class (BME 524), taken Spring 2026 at ASU from Dr. Chris Plaisier. 

# Course learning outcomes
## Students will be able to:
- Design omics studies that provide ample statistical power for downstream analyses
- Identify the correct quantitative systems biology approach to analyze omics datasets: clustering, classification, enrichment analysis, network construction, or trajectory analysis
- Use quantitative reasoning to interpret systems biology analyses
- Understand how to turn interpretations into experimental studies that validate hypotheses

# Module Topics and Skills
- Modules 1-10 were lectures with no coding component, covering these topics: systems biology, complex systems, reading primary literature, designing experiments, sequencing, genomics, epigenomics, transcriptomics, proteomics, and metabolomics. For each of the omics lectures, we covered the types of experiments are performed, the information they provide, and the advantages and disadvantages of each.
- Module 11: Defining cell types in scRNA-seq (clustering)
    1. Explain how scRNA-seq data are generated and processed, from FASTQ files to a gene Ã— cell count matrix.
    2. Perform quality control and preprocessing in Scanpy, including filtering low-quality cells, normalizing counts, and regressing out confounders.
    3. Conduct dimensionality reduction and clustering (PCA, UMAP, Leiden) to identify cellular subpopulations.
    4. Identify and interpret marker genes using differential expression analysis to assign biological identities to clusters.
    5. Apply reproducible Python workflows to analyze and visualize PBMC heterogeneity in a systems biology context.
- Module 12: 
- Module 13: 